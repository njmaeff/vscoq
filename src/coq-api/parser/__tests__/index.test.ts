import {Parser, trimLeft, ARROW, COLON, DEFINE, IDENTIFIER, PIPE, STOP} from "../"


describe('token tests', () =>
{
    test('identifier', () =>
    {
        let id = 'Inductive  '
        let match = IDENTIFIER.exec(id)?.[0];
        expect(match).toEqual('Inductive');
    });
    test('colon', () =>
    {
        let id = ':  '
        let match = COLON.exec(id)?.[0];
        expect(match).toEqual(':');
    });

    test('define', () =>
    {
        let id = ':=  '
        let match = DEFINE.exec(id)?.[0];
        expect(match).toEqual(':=');
    });
    test('stop', () =>
    {
        let id = '.  '
        let match = STOP.exec(id)?.[0];
        expect(match).toEqual('.');
    });

    test('pipe', () =>
    {
        let id = '|  '
        let match = PIPE.exec(id)?.[0];
        expect(match).toEqual('|');
    });

    test('arrow', () =>
    {
        let id = '->  '
        let match = ARROW.exec(id)?.[0];
        expect(match).toEqual('->');
    });

});


describe('trim left', () =>
{
    test('trim left', () =>
    {
        let str = `              \n\r                  \n    hello`
        let result = trimLeft(str);
        expect(result).toEqual({ column: 4, length: 39, row: 2, src: 'hello' });
    });

    test('no new space', () =>
    {
        let str = 'hello'
        let result = trimLeft(str);
        expect(result).toEqual({ column: 0, length: 0, row: 0, src: 'hello' })
    });

    test('one new line', () =>
    {
        let str = '\nhello'
        let result = trimLeft(str);
        expect(result).toEqual({ column: 0, length: 1, row: 1, src: 'hello' })
    });
});

describe('parser', () =>
{

    test('parse inductive type', () =>
    {
        let parser = Parser();
        let src = `Inductive nat : Set :=\n    | 0 : nat\n    | S : nat -> nat.`;
        let result = parser.parse(src)

        expect(result.ast).toEqual({
            type: 'program',
            start: {row: 0, column: 0},
            end: {row: 0, column: 60},
            nodes: [
                {
                    type: 'inductive',
                    identifier: {
                        name: 'nat',
                        start: {row: 0, column: 10},
                        end: {row: 0, column: 13},
                    },
                    source: src,
                    returns: 'Set',
                    definition: [
                        {
                            identifier: '0',
                            type: 'nat',
                            start: {row: 1, column: 6},
                            end: {row: 1, column: 7}
                        },
                        {
                            identifier: 'S',
                            type: 'nat -> nat',
                            start: {row: 2, column: 6},
                            end: {row: 2, column: 7}
                        },
                    ],
                    start: {row: 0, column: 0},
                    end: {row: 0, column: 60},
                }
            ]

        })
    });

});

describe('babel examples', () =>
{

    test('babel ast', () =>
    {
        const parser = require('@babel/parser');
        let src = `const x = () => "hello"`
        let ast = parser.parse(src)

        expect(ast).toMatchSnapshot(src);
    });
});
