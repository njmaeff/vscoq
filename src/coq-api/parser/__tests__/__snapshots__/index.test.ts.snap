// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel examples babel ast: const x = () => "hello" 1`] = `
Node {
  "comments": Array [],
  "end": 23,
  "errors": Array [],
  "loc": SourceLocation {
    "end": Position {
      "column": 23,
      "line": 1,
    },
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "program": Node {
    "body": Array [
      Node {
        "declarations": Array [
          Node {
            "end": 23,
            "id": Node {
              "end": 7,
              "loc": SourceLocation {
                "end": Position {
                  "column": 7,
                  "line": 1,
                },
                "identifierName": "x",
                "start": Position {
                  "column": 6,
                  "line": 1,
                },
              },
              "name": "x",
              "start": 6,
              "type": "Identifier",
            },
            "init": Node {
              "async": false,
              "body": Node {
                "end": 23,
                "extra": Object {
                  "raw": "\\"hello\\"",
                  "rawValue": "hello",
                },
                "loc": SourceLocation {
                  "end": Position {
                    "column": 23,
                    "line": 1,
                  },
                  "start": Position {
                    "column": 16,
                    "line": 1,
                  },
                },
                "start": 16,
                "type": "StringLiteral",
                "value": "hello",
              },
              "end": 23,
              "generator": false,
              "id": null,
              "loc": SourceLocation {
                "end": Position {
                  "column": 23,
                  "line": 1,
                },
                "start": Position {
                  "column": 10,
                  "line": 1,
                },
              },
              "params": Array [],
              "start": 10,
              "type": "ArrowFunctionExpression",
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 23,
                "line": 1,
              },
              "start": Position {
                "column": 6,
                "line": 1,
              },
            },
            "start": 6,
            "type": "VariableDeclarator",
          },
        ],
        "end": 23,
        "kind": "const",
        "loc": SourceLocation {
          "end": Position {
            "column": 23,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "start": 0,
        "type": "VariableDeclaration",
      },
    ],
    "directives": Array [],
    "end": 23,
    "interpreter": null,
    "loc": SourceLocation {
      "end": Position {
        "column": 23,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "script",
    "start": 0,
    "type": "Program",
  },
  "start": 0,
  "type": "File",
}
`;
